File: solutions/python/linked-list/linked-list-components.py
# LeetCode 817 — Linked List Components (Medium)
# Category: Linked List · Approach: Run counting
# Time: O(n + m) | Space: O(m)
# Source: https://leetcode.com/problems/linked-list-components/

def num_components(values: list[int], nums: list[int]) -> int:
    in_set = set(nums)
    count = 0
    in_run = False
    for v in values:
        if v in in_set:
            if not in_run:
                count += 1
                in_run = True
        else:
            in_run = False
    return count

File: solutions/typescript/linked-list/linked-list-components.ts
// LeetCode 817 — Linked List Components (Medium)
// Category: Linked List · Approach: Run counting
// Time: O(n + m) | Space: O(m)
// Source: https://leetcode.com/problems/linked-list-components/

function numComponents(values: number[], nums: number[]): number {
  const inSet = new Set(nums);
  let count = 0;
  let inRun = false;
  for (const v of values) {
    if (inSet.has(v)) {
      if (!inRun) {
        count += 1;
        inRun = true;
      }
    } else {
      inRun = false;
    }
  }
  return count;
}
